ref(sentry apps): Reorganize main integration platform page (#5011)
getsentry/sentry-docs: src/docs/product/integrations/integration-platform/public-integration.mdx

## Authorized Origins

It's possible to use auth tokens from the browser if you allow the origins of the pages making the requests. In the "Authorized JavaScript Origins" field, add each origin (for example, docs.sentry.io) separated by a newline. You don't need to include the protocol (http or https) in the origin. Currently, you can't use any wildcard characters (for example, `*.sentry.io`), so if you have multiple subdomains, you'll need to add them individually.

## Image Requirements

[Public integrations](/product/integrations/integration-platform/public-integration) require a logo image to be uploaded. The requirements for those are:
* Logo size must be between 256 &times; 256px and 1024 &times; 1024px
* File format must be PNG
* Background must be transparent (unless the logo takes up the entire space)

If the integration has a [UI component](/product/integrations/integration-platform/ui-components), you must also upload an icon that follows these requirements:
* Icon size must be between 256x256 px and 1024x1024px
* File must be a PNG in RGBA format

> * All pixels must be black or transparent

Reviewer: ```suggestion

* All pixels must be black or transparent
```

Typically, if you have the redirect URL configured, there is work happening on your end to "finalize" the installation. If this is the case, we recommend enabling the "Verify Install" option for your integration. Once enabled, you'll need to send a request marking the installation as officially "installed":

```python
requests.put(
    u'https://sentry.io/api/0/sentry-app-installations/{}/'.format(install_id),
    json={'status': 'installed'},
)
```

## Uninstallation

When a user uninstalls your integration, you'll receive an `installation.deleted` request to your webhook URL. For more information, check out the [full Webhooks documentation](/product/integrations/integration-platform/webhooks/).

## Authorized Origins

It's possible to use auth tokens from the browser if you allow the origins of the pages making the requests. In the "Authorized JavaScript Origins" field, add each origin (for example, docs.sentry.io) separated by a newline. You don't need to include the protocol (http or https) in the origin. Currently, you can't use any wildcard characters (for example, `*.sentry.io`), so if you have multiple subdomains, you'll need to add them individually.

> ## Image Requirements

Reviewer 2: Both of the sections (internal and public) for Image Requirements have this weird formatting thing where the first three bullets have no newline spaces after them, but the second list has a space between.

Not sure which we like better but I think we should stick to one style for them both.
![image](https://user-images.githubusercontent.com/35509934/167730707-f53989f1-4dc5-464d-a8b3-318b09547bf4.png)

---
title: Public Integrations
sidebar_order: 1
description: "Learn more about Sentry's public integrations."
---

> Public integrations are for the "general public" of Sentry users. These start in an unpublished state for development purposes and can be submitted later for approval to publish. For more information, check out the [section on Publication](#publication).

Reviewer: ```suggestion
Public integrations are for the "general public" of Sentry users. These start in an unpublished state for development purposes and can be submitted later for approval to publish. For more information, check out the [Publication section](#publication) of this page.
```

> The code examples in the sections below demonstrate a potential use-case that involves a Flask app receiving new issue webhooks from Sentry, calling the Sentry API for more data about the issue, and pushing it to Pushover as a generator of desktop/mobile notifications.

Reviewer: ```suggestion
The code examples in the sections below demonstrate a potential use case that involves a Flask app receiving new issue webhooks from Sentry, calling the Sentry API for more data about the issue, and pushing it to Pushover as a generator of desktop/mobile notifications.
```

However, if your integration has a redirect URL configured, the integration redirects the user’s browser to the configured URL with the grant code and installation ID in the query params.

Start building your integration by implementing the redirect URL endpoint, `/setup` — typically where you exchange the grant code for a token that allows you to make API calls to Sentry:

```python
import requests
from flask import redirect, request

@app.route('/setup', methods=['GET'])
def setup():
    code = request.args.get('code')
    install_id = request.args.get('installationId')

    url = u'https://sentry.io/api/0/sentry-app-installations/{}/authorizations/'
    url = url.format(install_id)

    payload = {
        'grant_type': 'authorization_code',
        'code': code,
        'client_id': 'your-client-id',
        'client_secret': 'your-client-secret',
    }

    resp = requests.post(url, json=payload)
    data = resp.json()

    token = data['token']
    refresh_token = data['refreshToken']
    # ... Securely store the install_id, token and refresh_token in DB ...

    return redirect('https://sentry.io/settings/')
```

## Auth Tokens

> Public integrations generate [Authentication Tokens](/product/integrations/integration-platform/#auth-tokens) after a successful [OAuth Installation](#oauth-process).

Reviewer: ```suggestion
Public integrations generate [authentication tokens](/product/integrations/integration-platform/#auth-tokens) after a successful [OAuth installation](#oauth-process).
```

These tokens automatically expire every eight hours, meaning they must be refreshed manually.

> 

Reviewer: ```suggestion
```

> ## Refreshing Tokens

Reviewer: Since this has to do with auth tokens, does it make sense to make this a sub-section of that?

```suggestion
### Refreshing Tokens
```

}
```

Tokens expire after eight hours, so you'll need to refresh your tokens accordingly:

```python
def refresh_token(install_id):
    url = u'https://sentry.io/api/0/sentry-app-installations/{}/authorizations/'
    url = url.format(install_id)

    refresh_token = retrieve_refresh_token_from_db(install_id)

    payload = {
        'grant_type': 'refresh_token',
        'refresh_token': refresh_token,
        'client_id': 'your-client-id',
        'client_secret': 'your-client-secret',
    }

    resp = requests.post(url, json=payload)
    data = resp.json()

    new_token = data['token']
    new_refresh_token = data['refreshToken']
    # ... Securely update the token and refresh_token in DB...

    return new_token
```

The data you can expect back for both the initial grant code exchange and subsequent token refreshes is as follows:

```json
{
  "id": "38",
  "token": "ec48bf98637d44c294ead7566513686237e74ab67a074c64b3aaca2d93dbb8f1",
  "refreshToken": "c866f154a65841638d44ee26364409b0a1a67bd642bd46e7a476f34f810712d6",
  "dateCreated": "2019-08-07T20:25:09.870Z",
  "expiresAt": "2019-08-08T04:25:09.870Z",
  "state": null,
  "application": null
}
```

> 

Reviewer: ```suggestion
```

## Verifying Installations

Typically, if you have the redirect URL configured, there is work happening on your end to "finalize" the installation. If this is the case, we recommend enabling the "Verify Install" option for your integration. Once enabled, you'll need to send a request marking the installation as officially "installed":

```python
requests.put(
    u'https://sentry.io/api/0/sentry-app-installations/{}/'.format(install_id),
    json={'status': 'installed'},
)
```

> 

Reviewer: ```suggestion
```

## Uninstallation

> When a user uninstalls your integration you'll receive an `installation.deleted` request to your webhook URL. For more information, check out the [full documentation on Webhooks](/product/integrations/integration-platform/webhooks/).

Reviewer: ```suggestion
When a user uninstalls your integration, you'll receive an `installation.deleted` request to your webhook URL. For more information, check out the [full Webhooks documentation](/product/integrations/integration-platform/webhooks/).
```

## Authorized Origins

It's possible to use auth tokens from the browser if you allow the origins of the pages making the requests. In the "Authorized JavaScript Origins" field, add each origin (for example, docs.sentry.io) separated by a newline. You don't need to include the protocol (http or https) in the origin. Currently, you can't use any wildcard characters (for example, `*.sentry.io`), so if you have multiple subdomains, you'll need to add them individually.

> 

Reviewer: ```suggestion
```

## Publication

When you're ready for the publication process, click "Publish" next to the integration you want to submit. This sends an email to partners@sentry.io, letting us know your integration is ready for review:

![Buttons providing the options to publish or delete.](publish-button.png)

### Image Requirements

[Public integrations](/product/integrations/integration-platform/public-integration) require a logo image to be uploaded. The requirements for those are:
* Logo size must be between 256 &times; 256px and 1024 &times; 1024px
* File format must be PNG
* Background must be transparent (unless the logo takes up the entire space)

If the integration has a [UI component](/product/integrations/integration-platform/ui-components), you must also upload an icon that follows these requirements:
* Icon size must be between 256x256 px and 1024x1024px
* File must be a PNG in RGBA format
* All pixels must be black or transparent

> Internal integrations can have logos and icons uploaded, but they are not required.

Reviewer: This no longer makes sense as part of this section.

```suggestion
```

---
title: Public Integrations
sidebar_order: 1

> description: "Learn more about Sentry's Public Integrations."

Reviewer: ```suggestion
description: "Learn more about Sentry's public integrations."
```

```python
url = u'https://sentry.io/api/0/sentry-app-installations/{}/authorizations/'
url = url.format(install_id)

payload = {
    'grant_type': 'authorization_code',
    'code': code,
    'client_id': 'your-client-id',
    'client_secret': 'your-client-secret',
}
```

Tokens expire after eight hours, so you'll need to refresh your tokens accordingly:

```python
url = u'https://sentry.io/api/0/sentry-app-installations/{}/authorizations/'
url = url.format(install_id)

refresh_token = retrieve_refresh_token_from_db(install_id)

payload = {
    'grant_type': 'refresh_token',
    'refresh_token': refresh_token,
    'client_id': 'your-client-id',
    'client_secret': 'your-client-secret',
}
```

The data you can expect back for both the initial grant code exchange and subsequent token refreshes is as follows:

```json
{
  "id": "38",
  "token": "ec48bf98637d44c294ead7566513686237e74ab67a074c64b3aaca2d93dbb8f1",
  "refreshToken": "c866f154a65841638d44ee26364409b0a1a67bd642bd46e7a476f34f810712d6",
  "dateCreated": "2019-08-07T20:25:09.870Z",
  "expiresAt": "2019-08-08T04:25:09.870Z",
  "state": null,
  "application": null
}
```

> ## Auth Tokens

Reviewer 2: As per my other comment on the Integration Platform page, I think we can reduce the duplicate stuff by changing this to something like:

> ## Authentication Tokens
> Public integrations generate [Authentication Tokens](/integration-platform#authentication-tokens) after a successful [OAuth Installation](#installation)
> These tokens automatically expire every eight hours, meaning they must be refreshed manually.

Auth Tokens are referred to as _tokens_ in the Integration Platform. Auth tokens are passed using an auth header and are used to authenticate as a user account with the API. In the case of the Integration Platform, these tokens are used to authenticate as an application (with access to a specific organization) instead of a user.

## Using Tokens for Requests

When making requests to the Sentry API, you use the token just like you would when you're making typical [API requests](/api/auth/). Tokens are associated with the installation, meaning they have access to the Sentry organization that installed your integration.

## Token Expiration

Tokens for public integrations expire every eight hours.

> ## Refreshing Tokens

Reviewer 2: After that you can probably get rid of this section all together. If it reads better, you can change the 'Token Exchange' heading to read 'Refreshing Tokens', but that's up to you.

        'client_secret': 'your-client-secret',
    }

    resp = requests.post(url, json=payload)
    data = resp.json()

    new_token = data['token']
    new_refresh_token = data['refreshToken']
    # ... Securely update the token and refresh_token in DB...

    return new_token
```

Instead of keeping track of times and passively refreshing when a token expires, another way you can handle refreshing tokens is to actively capture exceptions raised by requests that receive a 401 unauthorized response from Sentry, refresh the token, and remake the request.

### Verifying Installations (optional)

Typically, if you have the redirect URL configured, there is work happening on your end to "finalize" the installation. If this is the case, we recommend enabling the "Verify Install" option for your integration. Once enabled, you'll need to send a request marking the installation as officially "installed":

```python
requests.put(
    u'https://sentry.io/api/0/sentry-app-installations/{}/'.format(install_id),
    json={'status': 'installed'},
)
```

## Uninstallation

When a user uninstalls your integration, you'll receive a webhook request to your webhook URL.

> ## Integration Webhooks

Reviewer 2: I think this entire section can be removed as well. I think once the reference implementation is built we can just mention it here instead of trying to paste in big code blobs where the intention of the code isn't immediately obvious.

```python
@app.route('/webhook', methods=['POST'])
def webhook():
    data = json.loads(request.data)

    if data['action'] != 'created':
        return

    issue_id = data['data']['issue']['id']
    install_id = data['installation']['uuid']

    issue_details = get_sentry_issue(install_id, issue_id)

    if issue_details['project']['name'] != 'Backend':
        return

    event = data['data']['event']

    payload = {
        'user': 'pushover-user-key',
        'token': 'pushover-api-token',
        'message': event['message'][:1024],
        'title': event['message'][:250],
        'url': event['url'],
        'url_title': 'Issue Details',
        'priority': 0,
    }
    requests.post('https://api.pushover.net/1/messages.json', data=payload)

def get_sentry_issue(install_id, issue_id):
    token = retrieve_from_db(install_id)

    url = u'https://sentry.io/api/0/issues/{}/'.format(issue_id)
    headers = {'Authorization': u'Bearer {}'.format(token)}

    resp = requests.get(url, headers=headers)
    return resp.json()
```

For more information, check out the [full documentation on Webhooks](/product/integrations/integration-platform/webhooks/).

> ## Published State

Reviewer 2: I think this heading can be changed to 'Publishing' since that'd match what you reference at the top of the page. I also think this would be a good section to include the image requirements for publishing, maybe as an H3 underneath here.
